src/keymgmt.c

```diff
@@ -986,90 +1004,124 @@ const OSSL_DISPATCH p11prov_rsa_keymgmt_functions[] = {
-              ctx->data.rsa.allowed_types_size },
-        };
-        CK_ULONG tsize = 1;
+            key = p11prov_rsa_gen_internal(genctx, cb_fn, cb_arg, true);
+            if (!key) {
+                /* Failed: This may be because the token does not support
+                 * CKA_ALLOWED_MECHS, so we retry again later without */
+                P11PROV_debug("Failed to Generate PSS key with restrictions");
 
-        ret = p11prov_obj_set_attributes(ctx->provctx, NULL, key, template,
-                                         tsize);
-        if (ret != CKR_OK) {
-            P11PROV_debug("Failed to add RSAPSS restrictions (%lu)", ret);
-            if (ret == CKR_ATTRIBUTE_TYPE_INVALID) {
-                /* set quirk to disable future attempts for this token */
                 token_supports_allowed_mechs = CK_FALSE;
-                (void)p11prov_token_sup_attr(
-                    ctx->provctx, p11prov_obj_get_slotid(key), SET_ATTR,
+                /* if he quirk has never been set we'll get back what we
+                 * defaulted to, if that is the case then set the quirk.
+                 * Otherwise the quirk was successfully probed earlier
+                 * and we'll ignore setting anything as this may also be
+                 * a fluke or a key specific failure */
+                ret = p11prov_token_sup_attr(
+                    ctx->provctx, p11prov_obj_get_slotid(key), GET_ATTR,
                     CKA_ALLOWED_MECHANISMS, &token_supports_allowed_mechs);
+                if (ret != CKR_OK) {
+                    P11PROV_raise(ctx->provctx, ret, "Failed to probe quirk");
+                } else if (token_supports_allowed_mechs == CK_FALSE) {
+                    /* The previous check didn't hit a stored quirk, so
+                     * signal to set one later based on the outcome of the
+                     * next attempt. */
+                    set_quirk = true;
+                }
             }
         }
     }
 
-    ret = CKR_OK;
+    if (!key) {
+        key = p11prov_rsa_gen_internal(genctx, cb_fn, cb_arg, false);
+    }
 
-done:
-    if (ret != CKR_OK) {
-        p11prov_obj_free(key);
-        key = NULL;
+    if (key && set_quirk) {
+        token_supports_allowed_mechs = CK_FALSE;
+        (void)p11prov_token_sup_attr(ctx->provctx, p11prov_obj_get_slotid(key),
+                                     SET_ATTR, CKA_ALLOWED_MECHANISMS,
+                                     &token_supports_allowed_mechs);
     }
+
+done:
     return key;
 }
 
+static int p11prov_rsapss_gen_set_params(void *genctx,
+                                         const OSSL_PARAM params[])
+{
+    struct key_generator *ctx = (struct key_generator *)genctx;
+    const OSSL_PARAM *p;
+    int ret;
+
+    if (!ctx) {
+        return RET_OSSL_ERR;
+    }
+
+    if (params == NULL) {
+        return RET_OSSL_OK;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_DIGEST);
+    if (p) {
```

Comment:
Should we try to set all PSS mechanisms in case there is no digest specified here in the else branch as we did previously with `set_default_rsapss_mechanisms()`, which is now deleted?

Reply:
No, I deleted it on purpose because setting all mechanisms is not really useful, omitting the attribute completely is more efficient.

Reply:
This was setting just the PSS mechanisms technically preventing to use RAW rsa and PKCS#1.5, which was the main point, you will not be able to achieve without that (unless you force specific digest).

Reply:
I think that was main point of introducing RSA-PSS as separate key type in openssl was to not allow any other operations on these keys so that would be my preference to do this enforcing, if the token supports it.